dafny-lang/dafny
# Make code actions available in DafnyCore

### Changes
- Merge the classes `ErrorDetail` and `DafnyCodeActions` into `ErrorRegistry`, which is located in DafnyCore. As such, error descriptions and code actions can be registered from within DafnyCore. This allows:
  - Creating a command `dafny format --code-actions` which applies code actions in addition to formatting`
  - Registering code actions for errors in the same locations as where the errors are created.
- `DiagnosticDafnyCodeActionProvider.GetDafnyCodeActions` now takes a `DafnyDiagnostic` that has an `IToken`, which makes it so that no ranges have to be passed back and forth in the `Data` field of code actions.
- Added an alternative to `TestCodeActionHelper`, `TestCodeAction` that allows specifying the range in which the action replaces things, which was needed for one of the new tests. @MikaelMayer, it might be good to unify `TestCodeAction` and `TestCodeActionHelper`. Could you comment on the markup syntax that I used for `TestCodeAction`? It's a little ugly but it does reuse `MarkupTestFile`.
- Add `TODO` markers for various errors which are registered in `ParseErrors` but which are never reported by the Parser. This is effectively dead code. 

### Testing
- Add a test for the code action that removes an illegal underscore
- Add a test for the code action that removes a deprecated attribute

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>
